Check Trident version to detect compatibility mode

Darsstar opened the pull request:
IE11 in compatibility mode is actually not too old. It's just not recognized as IE11 if you leave the Trident version out of the equation.
Checking the Trident version is something Vaadin 7 (since 7.0.0.beta5) does: https://github.com/vaadin/framework/commit/fc15f166742737cabfa80e9d548e6d8a1a1f7a04
Trident 4 -> IE8
Trident 5 -> IE9
Trident 6 -> IE10
Trident 7 -> IE11 (windows 7, 10)
Trident 8 -> IE11 (windows 10 initial release only)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/8884)
<!-- Reviewable:end -->

Darsstar committed:
Check Trident version to detect compatibility mode

IE11 in compatibility mode is actually not too old. It's just not recognized as IE11 if you leave the Trident version out of the equation.
Checking the Trident version is something Vaadin 7 (since 7.0.0.beta5) does: https://github.com/vaadin/framework/commit/fc15f166742737cabfa80e9d548e6d8a1a1f7a04
Trident 4 -> IE8
Trident 5 -> IE9
Trident 6 -> IE10
Trident 7 -> IE11 (windows 7, 10)
Trident 8 -> IE11 (windows 10 initial release only)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/framework?pullRequest=8884) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/framework?pullRequest=8884) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/framework?pullRequest=8884) it.</sub>

Darsstar committed:
Derrive IE version from Trident version

Check the Trident version in a different place and use it to determine the actual IE version instead of trusting the possibly emulated MSIE version string.

pleku commented:
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8884#-KfjnqfYSy1yw2pF1id6:-KfjnqfYSy1yw2pF1id7:b7e9abw) (no related file):*
Hi! Thanks for the contribution. It would nice to add a test for this in `VBrowserDetailsUserAgentParserTest` so that it will not get removed again.

If you don't have the time to do that, I can add the test myself, but the main question is: what is the user agent for IE11 in compatibility mode ? Need to try to look that up

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8884)*
<!-- Sent from Reviewable.io -->

pleku commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[shared/src/main/java/com/vaadin/shared/VBrowserDetails.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8884#-Kfk_XBOZohlUNWRfZN1:-Kfk_XBPz9DyPbzomtqz:bgxz0jd) ([raw file](https://github.com/vaadin/framework/blob/3f40c31a8e63d78eafc5fe4b78bed6d6dac309b6/shared/src/main/java/com/vaadin/shared/VBrowserDetails.java#L155)):*
> ```Java
>                     int majorVersion = browserMajorVersion = 11;
>                     if (browserEngineVersion <= 7) {
>                         majorVersion = (int) (browserEngineVersion + 4);
> ```

This is the kind of magic that I should be described in a comment, I know Microsoft works in mysterious ways, but still, where does this come from ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8884)*
<!-- Sent from Reviewable.io -->

pleku commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8884)*
<!-- Sent from Reviewable.io -->

Darsstar committed:
Added IE11 user agent test cases

- IE11 Windows 7 compatibility view (CV) IE7
- IE11 Windows 10 CV IE7
- IE11 launch day Windows 10 CV IE7

Darsstar committed:
Refactored Trident version parsing and made 7 the upper limit

Darsstar committed:
Simplified the IE version detection a bit

Darsstar commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8884#-KfjnqfYSy1yw2pF1id6:-Kfl5ThEM0c0AkRJBs04:bdxr8p6) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Hi! Thanks for the contribution. It would nice to add a test for this in `VBrowserDetailsUserAgentParserTest` so that it will not get removed again.

If you don't have the time to do that, I can add the test myself, but the main question is: what is the user agent for IE11 in compatibility mode ? Need to try to look that up

</blockquote></details>

Added three new tests.

For reference:

When in compatibility mode (they all start reporting IE7 by default)
IE8: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; Trident/4.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C)`
IE9: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C)`
IE10: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; Trident/6.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C)`
IE11 - Win 7: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/7.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E)`
IE11 - Win 10: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 10.0; WOW64; Trident/7.0; .NET4.0C; .NET4.0E)`
IE11 - Win 10 launch day: `Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 10.0; WOW64; Trident/8.0; .NET4.0C; .NET4.0E)`

---

*[shared/src/main/java/com/vaadin/shared/VBrowserDetails.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8884#-Kfk_XBOZohlUNWRfZN1:-Kfl0iz12uiYuMFEy9dd:b-59n35h) ([raw file](https://github.com/vaadin/framework/blob/3f40c31a8e63d78eafc5fe4b78bed6d6dac309b6/shared/src/main/java/com/vaadin/shared/VBrowserDetails.java#L155)):*
> his is the kind of magic that I should be described in a comment, I know Microsoft works in mysterious ways, but still, where does this come from ?
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8884)*
<!-- Sent from Reviewable.io -->

pleku commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8884)*
<!-- Sent from Reviewable.io -->

Darsstar commented:
@pleku Will this end up in 8.0.4 or should I have approached making this contribution differently if that was my goal? And if so, what would I need to do to get it merged into 8.0.4 too?

pleku commented:
@Darsstar don't worry I'm picking it to 8.0.4
